wurst/systems/boards/TribeBoard.wurst

```diff
@@ -131,8 +189,25 @@ init
             Tribe.getTribes().each(tribe -> new TribeBoard(tribe))
 
         registerPlayerEvent(EVENT_PLAYER_LEAVE) ->
-            let leaver = EventData.getTriggerPlayer()
-            if Tribe.ofPlayer(leaver) != null
+            if not GetTriggerPlayer().isObserver() and not GetTriggerPlayer().isObserverNew()
                 TribeBoard
-                    .findForPlayer(leaver)
-                    .fixDisplay()
+                    .findForPlayer(GetTriggerPlayer())
+                    ..fixDisplay()
+                    ..updateLeaverName(GetTriggerPlayer())
+
+    EventListener.add(EVENT_PLAYER_HERO_LEVEL) ->
+        if GetTriggerUnit().isTroll()
+            let owner = GetTriggerUnit().getOwner()
+            TribeBoard
+                .findForPlayer(owner)
+                .setTrollLevel(owner)
+
+    registerAfterEffect() (unit target, int unitID) ->
+        if target.isTroll()
+            let owner = target.getOwner()
+            // For some reason, the troll level isn't updated in time with this event, but the name is
+            // waiting a bit to get the correct level
+            nullTimer() ->
```

Comment:
Does this not work with the null timer? In that case the transformation library should probably include the null timer.

Reply:
It works with the nulltimer.

Reply:
Oops, had a typo. I meant: does it work without the `nullTimer`? If you don't have the `nullTimer`, are values for the original troll used?

Reply:
Oh, nevermind. This is because the level is reset not by the transformation system. Right now it happens:

1. Transformation system changes the unit type and that's it.
2. It fires callbacks, including this.
3. The `TrollUpgrade.wurst` package changes the level post transformation.

We should either have a callback system for troll transformation specifically, move the level reset prior to the transformation, or switch to a single level track.